## Proyecto Pandas 
##### Jacobo Maté

In [1]:
import numpy as np # Importamos paquetes
import pandas as pd
import string

In [2]:
path = '../data/attacks.csv' # Importamos base de datos
data = pd.read_csv(path, encoding = "ISO-8859-1")

In [3]:
pd.set_option('display.max_columns', None)  # ver todas las columnas

pd.set_option('display.max_rows', None)     # ver todas las filas

In [4]:
data.head() # Mostramos la cabeza de los datos

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2  original order Unnamed: 22 Unnamed: 23  
0    2018.06.25          6303.0         NaN         NaN  
1    2018.06.18          6302.0         NaN         NaN  
2    2018.06.09          6301.0         NaN         NaN  
3    2018.06.08          6300.0         NaN         NaN  
4    2018.06.04          6299.0         NaN         NaN

#### ¿Qué podemos sacar de este head?
 * Tenemos varias columnas con título con mayúsculas y espacios, voy a cambiarlas todas a lower y hacer replace de espacios. Esto es útil de cara a más tarde poder usar punto en vez de corchetes y comillado al querer referirme a una columna

In [5]:
data.columns = [c.lower().replace(' ', '') for c in data.columns] # Paso a minúsculas y quito los espacios para poder usar puntos
data.columns

Index(['casenumber', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal(y/n)', 'time',
       'species', 'investigatororsource', 'pdf', 'hrefformula', 'href',
       'casenumber.1', 'casenumber.2', 'originalorder', 'unnamed:22',
       'unnamed:23'],
      dtype='object')

#### Limpieza | Ordenado datos `nans` y `duplicates`

In [6]:
data.info() # Procedemos a imprimir el tipo de datos con el que vamos a trabajar

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   casenumber            8702 non-null   object 
 1   date                  6302 non-null   object 
 2   year                  6300 non-null   float64
 3   type                  6298 non-null   object 
 4   country               6252 non-null   object 
 5   area                  5847 non-null   object 
 6   location              5762 non-null   object 
 7   activity              5758 non-null   object 
 8   name                  6092 non-null   object 
 9   sex                   5737 non-null   object 
 10  age                   3471 non-null   object 
 11  injury                6274 non-null   object 
 12  fatal(y/n)            5763 non-null   object 
 13  time                  2948 non-null   object 
 14  species               3464 non-null   object 
 15  investigatororsourc

In [7]:
data.shape

(25723, 24)

Tenemos un total de 25723 filas y 24 columnas

#### ¿Qué podemos sacar ya del tipo de datos con los que vamos a trabajar?
* Tenemos un total de 22 columnas de tipo objeto - Serán cadenas de caracteres
* Tenemos dos columnas de tipo float - columnas `year` y `originalorder`
* Es raro que un año sea de tipo float -> No va a haber años float tipo 2023.5


In [8]:
# Queremos cambiar el tipo de dato de year a int.
# Para ello primero comprobamos si hay algun nan con any() - Sacamos la suma de nulos que hay en cada columna
data.isnull().sum()

casenumber              17021
date                    19421
year                    19423
type                    19425
country                 19471
area                    19876
location                19961
activity                19965
name                    19631
sex                     19986
age                     22252
injury                  19449
fatal(y/n)              19960
time                    22775
species                 22259
investigatororsource    19438
pdf                     19421
hrefformula             19422
href                    19421
casenumber.1            19421
casenumber.2            19421
originalorder           19414
unnamed:22              25722
unnamed:23              25721
dtype: int64

* En este output podemos ver la gran cantidad de valores nulos que hay. A simple vista se ve que mínimo hay 17000 nulos por columna. Coincide que en varias columnas hay exactamente 19421 nulos. 
* Partiendo de la hipótesis de que esos 19421 datos nulos van a concidir en fila con los otros, voy a decidir borrar dichas filas. 
* Para ello voy a tomar como número de nulos mas restrictivo el más pequeño 17021 - columna `casenumber` - sacaré los índices de dichas filas y borraré del database sus filas.

In [9]:
indices_malos = data[data.casenumber.isna() == True].index # Aquí usamos un filtro para almacenar en la variable indices_malos los indices de las filas de la columna casenumber que sean nulos
indices_malos

Index([ 5488,  8702,  8703,  8704,  8705,  8706,  8707,  8708,  8709,  8710,
       ...
       25712, 25713, 25714, 25715, 25716, 25717, 25718, 25719, 25720, 25721],
      dtype='int64', length=17021)

* Voy a revisar individualmente el caso del índice 5488 porque el resto parece que son seguidas e intuyo que el resto de valores tb sean nulos para ese rango 8702-25721

In [10]:
data[data.index == 5488]

casenumber                  date    year      type country     area  \
5488        NaN  Reported 06-Sep-1905  1905.0  Provoked     USA  Florida   

                          location activity         name sex  age  \
5488  Fort Pierce, St Lucie County  Fishing  David Curry   M  NaN   

                                                 injury fatal(y/n) time  \
5488  Lacerations to leg from hooked shark PROVOKED ...          N  NaN   

     species            investigatororsource                     pdf  \
5488     NaN  Muncie Evening Press, 8/6/1913  1905.09.06.R-Curry.pdf   

                                            hrefformula  \
5488  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href  casenumber.1  \
5488  http://sharkattackfile.net/spreadsheets/pdf_di...  1905.09.06.R   

      casenumber.2  originalorder unnamed:22 unnamed:23  
5488  1905.09.06.R          815.0        NaN        NaN

In [11]:
# Como si que hay datos para este ataque me lo voy a quedar. 
# Sí que voy a borrar los datos de las filas correspondientes al rango 8702-25721.

data.drop([i for i in range(8702,25721)],axis=0).shape

(8704, 24)

In [12]:
# Antes de aplicar el cambio veo que esté haciéndolo bien imprimiendo el shape. Hemos bajado las filas a 8704 así que bien -> Procedo con el inplace=True
data.drop([i for i in range(8702,25722)],axis=0,inplace=True)
data.shape

(8703, 24)

In [13]:
data.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

In [14]:
data.isnull().sum()

casenumber                 1
date                    2401
year                    2403
type                    2405
country                 2451
area                    2856
location                2941
activity                2945
name                    2611
sex                     2966
age                     5232
injury                  2429
fatal(y/n)              2940
time                    5755
species                 5239
investigatororsource    2418
pdf                     2401
hrefformula             2402
href                    2401
casenumber.1            2401
casenumber.2            2401
originalorder           2394
unnamed:22              8702
unnamed:23              8701
dtype: int64

Volviendo al punto donde nos había extrañado que el tipo de `year` fuera float; si queremos cambiar el tipo de esta columna a int tendremos que cambiar los nans a un valor -> sino al aplicar el astype() nos saltará un error.

Para ello voy a proceder como antes: saco los indices de las filas donde se encuentren valores nulos en la columna `year` -> Exploro tabla para esos índices y si nada me indica lo contrario cambio ese nan a algo con lo que pueda trabajar

In [15]:
yearsnulos = data[data.year.isna() == True].index
yearsnulos

Index([  187,  6079,  6302,  6303,  6304,  6305,  6306,  6307,  6308,  6309,
       ...
        8693,  8694,  8695,  8696,  8697,  8698,  8699,  8700,  8701, 25722],
      dtype='int64', length=2403)

In [16]:
data.loc[yearsnulos].head(10)

casenumber                  date  year        type    country  \
187   2017.01.08.R  Reported 08-Jan-2017   NaN     Invalid  AUSTRALIA   
6079  1836.08.19.R  Reported 19-Aug-1836   NaN  Unprovoked    ENGLAND   
6302             0                   NaN   NaN         NaN        NaN   
6303             0                   NaN   NaN         NaN        NaN   
6304             0                   NaN   NaN         NaN        NaN   
6305             0                   NaN   NaN         NaN        NaN   
6306             0                   NaN   NaN         NaN        NaN   
6307             0                   NaN   NaN         NaN        NaN   
6308             0                   NaN   NaN         NaN        NaN   
6309             0                   NaN   NaN         NaN        NaN   

            area    location      activity          name  sex  age  \
187   Queensland         NaN  Spearfishing  Kerry Daniel    M   35   
6079  Cumberland  Whitehaven      Swimming         a boy    M  NaN   
6302         NaN         NaN           NaN           NaN  NaN  NaN   
6303         NaN         NaN           NaN           NaN  NaN  NaN   
6304         NaN         NaN           NaN           NaN  NaN  NaN   
6305         NaN         NaN           NaN           NaN  NaN  NaN   
6306         NaN         NaN           NaN           NaN  NaN  NaN   
6307         NaN         NaN           NaN           NaN  NaN  NaN   
6308         NaN         NaN           NaN           NaN  NaN  NaN   
6309         NaN         NaN           NaN           NaN  NaN  NaN   

                                      injury fatal(y/n) time     species  \
187   No attack, shark made a threat display        NaN  NaN  Bull shark   
6079                                   FATAL          Y  NaN         NaN   
6302                                     NaN        NaN  NaN         NaN   
6303                                     NaN        NaN  NaN         NaN   
6304                                     NaN        NaN  NaN         NaN   
6305                                     NaN        NaN  NaN         NaN   
6306                                     NaN        NaN  NaN         NaN   
6307                                     NaN        NaN  NaN         NaN   
6308                                     NaN        NaN  NaN         NaN   
6309                                     NaN        NaN  NaN         NaN   

        investigatororsource                           pdf  \
187   Liquid Vision 1/8/2017  2017.01.08.R-KerryDaniel.pdf   
6079          C. Moore, GSAF   1835.08.19.R-Whitehaven.pdf   
6302                     NaN                           NaN   
6303                     NaN                           NaN   
6304                     NaN                           NaN   
6305                     NaN                           NaN   
6306                     NaN                           NaN   
6307                     NaN                           NaN   
6308                     NaN                           NaN   
6309                     NaN                           NaN   

                                            hrefformula  \
187   http://sharkattackfile.net/spreadsheets/pdf_di...   
6079  http://sharkattackfile.net/spreadsheets/pdf_di...   
6302                                                NaN   
6303                                                NaN   
6304                                                NaN   
6305                                                NaN   
6306                                                NaN   
6307                                                NaN   
6308                                                NaN   
6309                                                NaN   

                                                   href  casenumber.1  \
187   http://sharkattackfile.net/spreadsheets/pdf_di...  2017.01.08.R   
6079  http://sharkattackfile.net/spreadsheets/pdf_di...  1836.08.19.R   
6302                                        

In [17]:
data.loc[yearsnulos].tail(10)

casenumber date  year type country area location activity name  sex  \
8693           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8694           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8695           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8696           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8697           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8698           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8699           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8700           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8701           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25722         xx  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       age injury fatal(y/n) time species investigatororsource  pdf  \
8693   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8694   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8695   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8696   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8697   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8698   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8699   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8700   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8701   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
25722  NaN    NaN        NaN  NaN     NaN                  NaN  NaN   

      hrefformula href casenumber.1 casenumber.2  originalorder unnamed:22  \
8693          NaN  NaN          NaN          NaN            NaN        NaN   
8694          NaN  NaN          NaN          NaN            NaN        NaN   
8695          NaN  NaN          NaN          NaN            NaN        NaN   
8696          NaN  NaN          NaN          NaN            NaN        NaN   
8697          NaN  NaN          NaN          NaN            NaN        NaN   
8698          NaN  NaN          NaN          NaN            NaN        NaN   
8699          NaN  NaN          NaN          NaN            NaN        NaN   
8700          NaN  NaN          NaN          NaN            NaN        NaN   
8701          NaN  NaN          NaN          NaN            NaN        NaN   
25722         NaN  NaN          NaN          NaN            NaN        NaN   

      unnamed:23  
8693         NaN  
8694         NaN  
8695         NaN  
8696         NaN  
8697         NaN  
8698         NaN  
8699         NaN  
8700         NaN  
8701         NaN  
25722        NaN

* Vista la salida, intuyo que desde la fila 6303 hasta 8701 tenemos valores nulos en todas las columnas salvo en `originalorder`.
* Como de 6302 a 8701 hay 2400 filas y sabemos que en total hay 2403 nulos, podemos confirmar que ese rango lo podemos eliminar todo también.
* Si todos los valores son nulos menos una columna tampoco me sirven de mucho -> Puedo hacer un último checkeo para comprobar que mi hipótesis sea verdadera y que no me esté fumando datos correctos por ahí enmedio.

In [18]:
data.drop([i for i in range(6303,8701)],axis=0).shape

(6305, 24)

In [19]:
data.drop([i for i in range(6303,8701)],axis=0, inplace=True)
data.shape

(6305, 24)

In [20]:
data.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

In [21]:
# Hacemos también un tail por si queda alguna fila nan al final del databse:
data.tail(10)

casenumber         date  year        type             country  \
6295     ND.0007  Before 1906   0.0  Unprovoked           AUSTRALIA   
6296     ND.0006  Before 1906   0.0  Unprovoked           AUSTRALIA   
6297     ND.0005  Before 1903   0.0  Unprovoked           AUSTRALIA   
6298     ND.0004  Before 1903   0.0  Unprovoked           AUSTRALIA   
6299     ND.0003    1900-1905   0.0  Unprovoked                 USA   
6300     ND.0002    1883-1889   0.0  Unprovoked              PANAMA   
6301     ND.0001    1845-1853   0.0  Unprovoked  CEYLON (SRI LANKA)   
6302           0          NaN   NaN         NaN                 NaN   
8701           0          NaN   NaN         NaN                 NaN   
25722         xx          NaN   NaN         NaN                 NaN   

                    area                             location      activity  \
6295                 NaN                                  NaN       Fishing   
6296     New South Wales                                           Swimming   
6297   Western Australia                          Roebuck Bay        Diving   
6298   Western Australia                                  NaN  Pearl diving   
6299      North Carolina                       Ocracoke Inlet      Swimming   
6300                 NaN                 Panama Bay 8ºN, 79ºW           NaN   
6301    Eastern Province  Below the English fort, Trincomalee      Swimming   
6302                 NaN                                  NaN           NaN   
8701                 NaN                                  NaN           NaN   
25722                NaN                                  NaN           NaN   

                        name  sex  age  \
6295               fisherman    M  NaN   
6296                Arab boy    M  NaN   
6297                    male    M  NaN   
6298                   Ahmun    M  NaN   
6299   Coast Guard personnel    M  NaN   
6300         Jules Patterson    M  NaN   
6301                    male    M   15   
6302                     NaN  NaN  NaN   
8701                     NaN  NaN  NaN   
25722                    NaN  NaN  NaN   

                                                  injury fatal(y/n) time  \
6295                                               FATAL          Y  NaN   
6296                                               FATAL          Y  NaN   
6297                                               FATAL          Y  NaN   
6298                                               FATAL          Y  NaN   
6299                                               FATAL          Y  NaN   
6300                                               FATAL          Y  NaN   
6301   FATAL. "Shark bit him in half, carrying away t...          Y  NaN   
6302                                                 NaN        NaN  NaN   
8701                                                 NaN        NaN  NaN   
25722                                                NaN        NaN  NaN   

                                                 species  \
6295                                       Blue pointers   
6296   Said to involve a grey nurse shark that leapt ...   
6297                                                 NaN   
6298                                                 NaN   
6299                                                 NaN   
6300                                                 NaN   
6301                                                 NaN   
6302                                                 NaN   
8701                                                 NaN   
25722                                                NaN   

                                    investigatororsource  \
6295   NY Sun, 9/9/1906, referring to account by Loui...   
6296   L. Becke in New York Sun, 9/9/1906; L. Schultz...   
6297                    H. Taunton; N. Bartlett,  p. 234   
6298               H. Taunton; N. Bartlett,  pp. 233-234   
6299                F. Schwartz, p.23; C. Creswell, GSAF   
6300                             

In [22]:
# Borramos también esas tres filas de nans enteras:
data.drop([6302,8701,25722],axis=0, inplace=True)

In [23]:
data.shape

(6302, 24)

In [24]:
data.isnull().sum()

casenumber                 1
date                       0
year                       2
type                       4
country                   50
area                     455
location                 540
activity                 544
name                     210
sex                      565
age                     2831
injury                    28
fatal(y/n)               539
time                    3354
species                 2838
investigatororsource      17
pdf                        0
hrefformula                1
href                       0
casenumber.1               0
casenumber.2               0
originalorder              0
unnamed:22              6301
unnamed:23              6300
dtype: int64

* Después de borrar los nans del paso anterior vuelvo a mostrar la suma de estos.
* A partir de aquí no vamos a quitar más filas xq entendemos que los valores nulos serán por falta de información - no siempre se sabrá especie de tiburón, edad, actividad que estaba practicando la víctima...
* Ahora todo lo que sea nan lo vamos a pasar a unknown utilizando el fillna()

In [25]:
data.fillna('unknown').tail()

casenumber         date year        type             country  \
6297    ND.0005  Before 1903  0.0  Unprovoked           AUSTRALIA   
6298    ND.0004  Before 1903  0.0  Unprovoked           AUSTRALIA   
6299    ND.0003    1900-1905  0.0  Unprovoked                 USA   
6300    ND.0002    1883-1889  0.0  Unprovoked              PANAMA   
6301    ND.0001    1845-1853  0.0  Unprovoked  CEYLON (SRI LANKA)   

                   area                             location      activity  \
6297  Western Australia                          Roebuck Bay        Diving   
6298  Western Australia                              unknown  Pearl diving   
6299     North Carolina                       Ocracoke Inlet      Swimming   
6300            unknown                 Panama Bay 8ºN, 79ºW       unknown   
6301   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       name sex      age  \
6297                   male   M  unknown   
6298                  Ahmun   M  unknown   
6299  Coast Guard personnel   M  unknown   
6300        Jules Patterson   M  unknown   
6301                   male   M       15   

                                                 injury fatal(y/n)     time  \
6297                                              FATAL          Y  unknown   
6298                                              FATAL          Y  unknown   
6299                                              FATAL          Y  unknown   
6300                                              FATAL          Y  unknown   
6301  FATAL. "Shark bit him in half, carrying away t...          Y  unknown   

      species                   investigatororsource  \
6297  unknown       H. Taunton; N. Bartlett,  p. 234   
6298  unknown  H. Taunton; N. Bartlett,  pp. 233-234   
6299  unknown   F. Schwartz, p.23; C. Creswell, GSAF   
6300  unknown                    The Sun, 10/20/1938   
6301  unknown                             S.W. Baker   

                                 pdf  \
6297          ND-0005-RoebuckBay.pdf   
6298               ND-0004-Ahmun.pdf   
6299  ND-0003-Ocracoke_1900-1905.pdf   
6300      ND-0002-JulesPatterson.pdf   
6301              ND-0001-Ceylon.pdf   

                                            hrefformula  \
6297  http://sharkattackfile.net/spreadsheets/pdf_di...   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...   
6299  http://sharkattackfile.net/spreadsheets/pdf_di...   
6300  http://sharkattackfile.net/spreadsheets/pdf_di...   
6301  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href casenumber.1  \
6297  http://sharkattackfile.net/spreadsheets/pdf_di...      ND.0005   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...      ND.0004   
6299  http://sharkattackfile.net/spreadsheets/pdf_di...      ND.0003   
6300  http://sharkattackfile.net/spreadsheets/pdf_di...      ND.0002   
6301  http://sharkattackfile.net/spreadsheets/pdf_di...      ND.0001   

     casenumber.2  originalorder unnamed:22 unnamed:23  
6297      ND.0005            6.0    unknown    unknown  
6298      ND.0004            5.0    unknown    unknown  
6299      ND.0003            4.0    unknown    unknown  
6300      ND.0002            3.0    unknown    unknown  
6301      ND.0001            2.0    unknown    unknown

* Nos falta hacer comprobar si hay duplicados - procedemos a ello

In [26]:
data.drop_duplicates(inplace=True)

In [27]:
data.shape

(6302, 24)

Tenemos la misma dimensión de filas por lo que asumimos que no había duplicados.

In [28]:
data.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

In [29]:
# Hago una copia de la base de datos para tener un chackpoint al que poder volver:
dc = data.copy() # dc por data clean

In [30]:
dc.shape

(6302, 24)

In [31]:
dc.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

#### Limpieza | Ordenado columnas `type` y `activity`

* Voy a continuar el estudio por obteniendo los valores únicos de la columna type. Ya que no se incluye una descripción de estos datos en la fuente de la base de datos vamos a ver qué tipos de detalles se incluyen en esta columna y si estos difieren mucho de un suceso a otro.

In [32]:
dc.type.unique()

array(['Boating', 'Unprovoked', 'Invalid', 'Provoked', 'Questionable',
       'Sea Disaster', nan, 'Boat', 'Boatomg'], dtype=object)

In [33]:
dc.type.value_counts()

type
Unprovoked      4595
Provoked         574
Invalid          547
Sea Disaster     239
Boating          203
Boat             137
Questionable       2
Boatomg            1
Name: count, dtype: int64

In [34]:
# Vamos a cambiar la de tipo 'Boatomg' por tipo Boating ya que creo se trata de una errata
ind_errata = dc[dc.type == 'Boatomg'].index
ind_errata

Index([1152], dtype='int64')

In [35]:
# Cambiamos el dato de type del indice 1152 por Boating
dc.type[dc.index==1152] = 'Boating'


C:\Users\jacom\AppData\Local\Temp\ipykernel_33004\1426339824.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dc.type[dc.index==1152] = 'Boating'


In [36]:
dc[dc.index == 1152]

casenumber         date    year     type      country          area  \
1152  2009.03.01.b  01-Mar-2009  2009.0  Boating  NEW ZEALAND  North Island   

      location activity                                             name sex  \
1152  Taranaki  Fishing  boat, occupants: Boyd Rutherford & Hamish Roper   M   

      age                                       injury fatal(y/n) time  \
1152  NaN  No injury to occupants, shark bit propeller          N  NaN   

     species           investigatororsource                      pdf  \
1152     NaN  Taranaki Daily News, 3/3/2009  2009.03.01-Taranaki.pdf   

                                            hrefformula  \
1152  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href  casenumber.1  \
1152  http://sharkattackfile.net/spreadsheets/pdf_di...  2009.03.01.b   

      casenumber.2  originalorder unnamed:22 unnamed:23  
1152  2009.03.01.b         5151.0        NaN        NaN

In [37]:
dc.type.value_counts()

type
Unprovoked      4595
Provoked         574
Invalid          547
Sea Disaster     239
Boating          204
Boat             137
Questionable       2
Name: count, dtype: int64

In [38]:
# También vamos a meter en una sola categoría los datos de boating y boat -> Boating
ind_boat = dc[dc.type == 'Boat'].index
ind_boat

Index([ 538,  565,  706,  818,  864,  956,  971, 1164, 1168, 1174,
       ...
       4188, 4214, 4242, 6070, 6087, 6127, 6136, 6145, 6212, 6274],
      dtype='int64', length=137)

In [39]:
# Cambiamos el dato de type del indice 1152 por Boating
dc.type.loc[ind_boat] = 'Boating'

C:\Users\jacom\AppData\Local\Temp\ipykernel_33004\4105221077.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dc.type.loc[ind_boat] = 'Boating'


In [40]:
dc.type.value_counts()

type
Unprovoked      4595
Provoked         574
Invalid          547
Boating          341
Sea Disaster     239
Questionable       2
Name: count, dtype: int64

In [41]:
dc.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

+ Dado que es la víctima la que decide si el ataque ha sido provocado o no, en mi opinión es un valor que no debería tenerse en cuenta ya que me parece totalmente sesgado.
* Seguimos con la actividad que practicaba la víctima del ataque

In [42]:
actividades = pd.DataFrame(dc.activity.value_counts())

In [43]:
actividades.shape

(1532, 1)

In [44]:
actividades.head(50)

count
activity                             
Surfing                           971
Swimming                          869
Fishing                           431
Spearfishing                      333
Bathing                           162
Wading                            149
Diving                            127
Standing                           99
Snorkeling                         89
Scuba diving                       76
Body boarding                      61
Body surfing                       49
Swimming                           47
Kayaking                           33
Treading water                     32
Pearl diving                       32
Fell overboard                     32
Free diving                        29
Boogie boarding                    29
Windsurfing                        19
Walking                            17
Boogie Boarding                    16
Shark fishing                      15
Floating                           14
Fishing                            13
Canoeing                           13
Surf skiing                        12
Surf fishing                       12
Surf-skiing                        12
Rowing                             12
Fishing for sharks                 11
Kayak Fishing                      11
Scuba Diving                       10
Freediving                         10
Sponge diving                      10
Fell into the water                 9
Sailing                             9
Sitting on surfboard                9
Diving for trochus                  9
Paddle boarding                     8
Sea disaster                        8
Diving for abalone                  7
Free diving for abalone             7
Skindiving                          7
Playing                             7
Surf skiing                         7
Surfing (sitting on his board)      7
Boating                             7
Floating on his back                7
Spearfishing                        7

Podemos ver que hay 1532 actividades diferentes - sin embargo muchas de ellas se podrian abarcar bajo campos parecidos

Algunos grandes campos pueden ser los siguientes:
* Fishing
* Swimming
* Surfing (any kind of board sport)
* Diving

Voy a llevar a cabo un estudio y actualizado de campos utilizando regex. 

In [45]:
import re # Importo paquete regex

# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_swim(cadena):

    patron = r'swim' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena, re.IGNORECASE)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Swimming'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

In [46]:
res = busca_swim('SWImidning in da pool') # Compruebo que funcione
print(res)

Swimming


In [47]:
res = busca_swim('cadena sin palabra')
print(res)

cadena sin palabra


In [48]:
# Ahora voy a hacer esta función exactamente igual para los otros casos de diving, fishing y surfing

In [49]:
# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_dive(cadena):

    cadena = cadena.lower()

    patron = r'div' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Diving'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

In [50]:
# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_fish(cadena):

    cadena = cadena.lower()

    patron = r'fish' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Fishing'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

In [51]:
# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_surf(cadena):

    cadena = cadena.lower()

    patron = r'surf' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Surfing'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

In [52]:
# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_over(cadena):

    cadena = cadena.lower()

    patron = r'over' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Overboard or capsized'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

In [53]:
# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_board(cadena): # Las actividades que sean de tipo board las voy a meter en surfing también

    cadena = cadena.lower()

    patron = r'board' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Surfing'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

Procedo a aplicar las 4 fórmulas que he definido a la columna `activity` con el formato apply

In [54]:
dc.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6302 entries, 0 to 6301
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   casenumber            6301 non-null   object 
 1   date                  6302 non-null   object 
 2   year                  6300 non-null   float64
 3   type                  6298 non-null   object 
 4   country               6252 non-null   object 
 5   area                  5847 non-null   object 
 6   location              5762 non-null   object 
 7   activity              5758 non-null   object 
 8   name                  6092 non-null   object 
 9   sex                   5737 non-null   object 
 10  age                   3471 non-null   object 
 11  injury                6274 non-null   object 
 12  fatal(y/n)            5763 non-null   object 
 13  time                  2948 non-null   object 
 14  species               3464 non-null   object 
 15  investigatororsource  6285

In [55]:
dc.activity = dc.activity.astype(str)

In [56]:
# Procedo a aplicar las 4 fórmulas a la columna activity del dataframe
dc.activity = dc.activity.apply(busca_swim)
dc.activity = dc.activity.apply(busca_dive)
dc.activity = dc.activity.apply(busca_surf)
dc.activity = dc.activity.apply(busca_fish)
dc.activity = dc.activity.apply(busca_over) # Llamo primero a esta función para quitarme los overboard antes de los deportes de tabla
dc.activity = dc.activity.apply(busca_board)

In [57]:
actividades_limpia = pd.DataFrame(dc.activity.value_counts())
actividades_limpia.head(15)

count
activity             
surfing          1219
fishing          1125
swimming         1123
diving            600
nan               544
Surfing           279
bathing           162
wading            149
standing           99
snorkeling         89
kayaking           33
treading water     32
walking            17
floating           14
canoeing           14

In [58]:
# He visto que hay 2 grupos de surfing -> Voy a pasar el que tiene mayúsculas a minúsculas para tenerlos juntos
dc.activity = dc.activity.str.replace('Surfing','surfing')

In [59]:
actividades_new = pd.DataFrame(dc.activity.value_counts())
actividades_new.shape

(641, 1)

Hemos conseguido bajar el tipo de actividades a 1/3 de la cantidad original.

In [60]:
dc.head(10)

casenumber         date    year        type    country             area  \
0    2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1    2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2    2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3    2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4    2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   
5  2018.06.03.b  03-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
6  2018.06.03.a  03-Jun-2018  2018.0  Unprovoked     BRAZIL       Pernambuco   
7    2018.05.27  27-May-2018  2018.0  Unprovoked        USA          Florida   
8  2018.05.26.b  26-May-2018  2018.0  Unprovoked        USA          Florida   
9  2018.05.26.a  26-May-2018  2018.0  Unprovoked        USA          Florida   

                                            location  activity  \
0                        Oceanside, San Diego County  paddling   
1                     St. Simon Island, Glynn County  standing   
2                                       Habush, Oahu   surfing   
3                                 Arrawarra Headland   surfing   
4                                           La Ticla    diving   
5                                 Flat Rock, Ballina   surfing   
6                              Piedade Beach, Recife  swimming   
7  Lighhouse Point Park, Ponce Inlet, Volusia County   fishing   
8                       Cocoa Beach, Brevard  County   walking   
9                      Daytona Beach, Volusia County  standing   

                     name sex  age  \
0             Julie Wolfe   F   57   
1         Adyson McNeely    F   11   
2             John Denges   M   48   
3                    male   M  NaN   
4          Gustavo Ramos    M  NaN   
5                 Chris 
   M  NaN   
6  Jose Ernesto da Silva    M   18   
7                    male   M   52   
8               Cody High   M   15   
9                    male   M   12   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   
5                            No injury, board bitten          N   
6                                              FATAL          Y   
7            Minor injury to foot. PROVOKED INCIDENT          N   
8                              Lower left leg bitten          N   
9                               Minor injury to foot          N   

             time          species              investigatororsource  \
0           18h00      White shark                  R. Collier, GSAF   
1   14h00  -15h00              NaN    K.McMurray, TrackingSharks.com   
2           07h45              NaN    K.McMurray, TrackingSharks.com   
3             NaN        2 m shark                    B. Myatt, GSAF   
4             NaN  Tiger shark, 3m                         A .Kipper   
5             NaN              NaN         Daily Telegraph, 6/4/2018   
6  Late afternoon      Tiger shark    Diario de Pernambuco, 6/4/2018   
7             NaN  Lemon shark, 3'   K. McMurray, TrackingSharks.com   
8           17h00   Bull shark, 6'    K.McMurray, TrackingSharks.com   
9           14h00              NaN  K. McMurray, Tracking Sharks.com   

                             pdf  \
0           2018.06.25-Wolfe.pdf   
1         2018.06.18-McNeely.pdf   
2          2018.06.09-Denges.pdf   
3       2018.06.08-Arrawarra.pdf   
4           2018.06.04-Ramos.pdf   
5      2018.06.03.b-FlatRock.pdf   
6       2018.06.03.a-daSilva.pdf   
7           2018.05.27-Ponce.pdf   
8          2018.05.26.b-High.pdf   
9  2018.05.26.a-DaytonaBeach.pdf   

                                         hrefformu

#### Limpieza columna `species`
Vamos a cambiar los valores nan de la columna por unkonwn

In [110]:
dc.species[dc.species.isna()].head() # Empezamos mostrando los valores nulos que hay en la columna

Series([], Name: species, dtype: object)

In [62]:
# Podemos comprobar que hay bastante valor nulo, es natural que no todas las víctimas de ataque desconozcan las especies de tiburón
dc.species.fillna('unknown', inplace=True)

In [63]:
dc.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location  activity             name sex  age  \
0     Oceanside, San Diego County  paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County  standing  Adyson McNeely    F   11   
2                    Habush, Oahu   surfing      John Denges   M   48   
3              Arrawarra Headland   surfing             male   M  NaN   
4                        La Ticla    diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00          unknown  K.McMurray, TrackingSharks.com   
2          07h45          unknown  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

#### Limpieza columna `country`

In [64]:
dc.country.unique()

array(['USA', 'AUSTRALIA', 'MEXICO', 'BRAZIL', 'ENGLAND', 'SOUTH AFRICA',
       'THAILAND', 'COSTA RICA', 'MALDIVES', 'BAHAMAS', 'NEW CALEDONIA',
       'ECUADOR', 'MALAYSIA', 'LIBYA', nan, 'CUBA', 'MAURITIUS',
       'NEW ZEALAND', 'SPAIN', 'SAMOA', 'SOLOMON ISLANDS', 'JAPAN',
       'EGYPT', 'ST HELENA, British overseas territory', 'COMOROS',
       'REUNION', 'FRENCH POLYNESIA', 'UNITED KINGDOM',
       'UNITED ARAB EMIRATES', 'PHILIPPINES', 'INDONESIA', 'CHINA',
       'COLUMBIA', 'CAPE VERDE', 'Fiji', 'DOMINICAN REPUBLIC',
       'CAYMAN ISLANDS', 'ARUBA', 'MOZAMBIQUE', 'FIJI', 'PUERTO RICO',
       'ITALY', 'ATLANTIC OCEAN', 'GREECE', 'ST. MARTIN', 'FRANCE',
       'PAPUA NEW GUINEA', 'TRINIDAD & TOBAGO', 'KIRIBATI', 'ISRAEL',
       'DIEGO GARCIA', 'TAIWAN', 'JAMAICA', 'PALESTINIAN TERRITORIES',
       'GUAM', 'SEYCHELLES', 'BELIZE', 'NIGERIA', 'TONGA', 'SCOTLAND',
       'CANADA', 'CROATIA', 'SAUDI ARABIA', 'CHILE', 'ANTIGUA', 'KENYA',
       'RUSSIA', 'TURKS & CAICOS', 'UNITE

In [65]:
dc.country.unique().shape

(213,)

A primera vista hemos mostrado la lista de valores únicos tiene 213 elementos. Esto de primeras nos debe extrañar ya que (a día de hoy) tan solo hay 206 países en el mundo. Esto se puede deber a que, tal y como se ve en la primera casilla ejecutada de este apartado podemos apreciar como algunos de los otputs no son países o como en la mayoría de los casos se escribe el nombre del país en mayúsculas sin embargo se ve alguna salida escrita en minúsculas. Vamos a proceder a arreglar esto.
Otro caso también presente en el nombre de los países es que se incluyen mares y oceanos. Este caso no se va a tocar ya que no podemos inventarnos un país ni tampoco eliminar dichos casos.

In [111]:
dc.country.value_counts().head()

country
USA             1913
AUSTRALIA       1093
SOUTH AFRICA     496
BRAZIL           100
NEW ZEALAND       89
Name: count, dtype: int64

In [67]:
dc.country[dc.country.isna()].shape

(50,)

In [68]:
# Cambiamos los nulos a 'unknown'
dc.country.fillna('unknown', inplace=True)

In [69]:
# Vamos a cambiar todas las palabras de la columna de country a mayúsculas:
dc.country = dc.country.str.upper()

In [70]:
dc.country.unique().shape # No parece habernos quitado más que 3 países

(210,)

#### Estudio columna `investigatororsource`

In [71]:
ind_sininvestigador = dc.investigatororsource[dc.investigatororsource.isna()].index # Creamos una lista de índices con los sucesos que tienen un NaN

In [72]:
dc.loc[ind_sininvestigador]

casenumber                  date    year        type       country  \
119    2017.07.14.a           14-Jul-2017  2017.0  Unprovoked     AUSTRALIA   
290    2016.04.07.b           07-Apr-2016  2016.0  Unprovoked           USA   
382      2015.07.27           27-Jul-2015  2015.0         NaN     AUSTRALIA   
631    2013.08.31.b           31-Aug-2013  2013.0  Unprovoked       BAHAMAS   
1130     2009.04.28           28-Apr-2009  2009.0  Unprovoked           USA   
1242   2008.06.28.a           28-Jun-2008  2008.0  Unprovoked  SOUTH AFRICA   
2125     1999.04.22           22-Apr-1999  1999.0  Unprovoked     MAURITIUS   
2144   1999.00.00.a                  1999  1999.0     Invalid           USA   
2460   1994.00.00.a                  1994  1994.0  Unprovoked           USA   
2648     1990.00.00                  1990  1990.0  Unprovoked           USA   
3508   1967.09.07.a           07-Sep-1967  1967.0  Unprovoked           USA   
3691     1964.03.00              Mar-1964  1964.0  Unprovoked       SENEGAL   
4058     1959.09.27           27-Sep-1959  1959.0     Invalid       BERMUDA   
4101  1959.04.09..a           09-Apr-1959  1959.0  Unprovoked  SOUTH AFRICA   
5321     1915.08.03           03-Aug-1915  1915.0  Unprovoked           USA   
6108   1823.12.30.R  Reported 30-Dec-1823  1823.0     Invalid       JAMAICA   
6147     1767.00.00                  1767  1767.0     Invalid        FRANCE   

                       area  \
119       Western Australia   
290                 Florida   
382                Victoria   
631                     NaN   
1130                Florida   
1242  Western Cape Province   
2125             Grand Baie   
2144               Virginia   
2460                Florida   
2648                Florida   
3508                Florida   
3691     Cap Vert Peninsula   
4058               Hamilton   
4101  Eastern Cape Province   
5321                Florida   
6108                    NaN   
6147           Côte d'Azur    

                                               location  activity  \
119                                       Floreat Beach   surfing   
290                         Florida Keys, Monroe County   fishing   
382                                    Tyrendarra Beach   surfing   
631                                      Freetown Beach   fishing   
1130                   St. Augustine, St. John's County       nan   
1242                                         Mossel Bay   surfing   
2125                              Pointe aux Canonniers  swimming   
2144  Sandridge Beach, Virginia Beach, Princess Anne...   surfing   
2460                      Flagler Beach, Flagler County   surfing   
2648                         Pensacola, Escambia County   surfing   
3508                            Key West, Monroe County   fishing   
3691                                     Bel Air, Dakar    diving   
4058                                     Astwood's Cove       nan   
4101                                    Port Elizabeth    fishing   
5321                     St. Augustine, St Johns County       nan   
6108                                                NaN       nan   
6147                                         St. Tropez   bathing   

                           name  sex  age  \
119                        male    M  NaN   
290             Jonathan Lester    M   34   
382                        male    M   40   
631               Bryan Collins    M  NaN   
1130                     Alicia    F  NaN   
1242               Kobus Maritz    M   46   
2125              Sylvia Lanner    F   41   
2144                       male    M  NaN   
2460               Jeff Weakley    M  NaN   
2648                       male    M   17   
3508                        NaN  NaN  NaN   
3691                       O.D.    M  NaN   
4058  Dorothy Barbara Rawlinson    F   29   
4101                       male    M  NaN   
5321            M.F. Turnipseed    M  NaN   
6108                       male  NaN  NaN   
6147            Sam

No vamos a eliminar los datos porque parecen reales, rellenaremos la columna con unknowns

In [73]:
dc.investigatororsource.fillna('unkown', inplace=True)

In [74]:
dc.loc[ind_sininvestigador] # Comprobamos haberlo hecho correctamente

casenumber                  date    year        type       country  \
119    2017.07.14.a           14-Jul-2017  2017.0  Unprovoked     AUSTRALIA   
290    2016.04.07.b           07-Apr-2016  2016.0  Unprovoked           USA   
382      2015.07.27           27-Jul-2015  2015.0         NaN     AUSTRALIA   
631    2013.08.31.b           31-Aug-2013  2013.0  Unprovoked       BAHAMAS   
1130     2009.04.28           28-Apr-2009  2009.0  Unprovoked           USA   
1242   2008.06.28.a           28-Jun-2008  2008.0  Unprovoked  SOUTH AFRICA   
2125     1999.04.22           22-Apr-1999  1999.0  Unprovoked     MAURITIUS   
2144   1999.00.00.a                  1999  1999.0     Invalid           USA   
2460   1994.00.00.a                  1994  1994.0  Unprovoked           USA   
2648     1990.00.00                  1990  1990.0  Unprovoked           USA   
3508   1967.09.07.a           07-Sep-1967  1967.0  Unprovoked           USA   
3691     1964.03.00              Mar-1964  1964.0  Unprovoked       SENEGAL   
4058     1959.09.27           27-Sep-1959  1959.0     Invalid       BERMUDA   
4101  1959.04.09..a           09-Apr-1959  1959.0  Unprovoked  SOUTH AFRICA   
5321     1915.08.03           03-Aug-1915  1915.0  Unprovoked           USA   
6108   1823.12.30.R  Reported 30-Dec-1823  1823.0     Invalid       JAMAICA   
6147     1767.00.00                  1767  1767.0     Invalid        FRANCE   

                       area  \
119       Western Australia   
290                 Florida   
382                Victoria   
631                     NaN   
1130                Florida   
1242  Western Cape Province   
2125             Grand Baie   
2144               Virginia   
2460                Florida   
2648                Florida   
3508                Florida   
3691     Cap Vert Peninsula   
4058               Hamilton   
4101  Eastern Cape Province   
5321                Florida   
6108                    NaN   
6147           Côte d'Azur    

                                               location  activity  \
119                                       Floreat Beach   surfing   
290                         Florida Keys, Monroe County   fishing   
382                                    Tyrendarra Beach   surfing   
631                                      Freetown Beach   fishing   
1130                   St. Augustine, St. John's County       nan   
1242                                         Mossel Bay   surfing   
2125                              Pointe aux Canonniers  swimming   
2144  Sandridge Beach, Virginia Beach, Princess Anne...   surfing   
2460                      Flagler Beach, Flagler County   surfing   
2648                         Pensacola, Escambia County   surfing   
3508                            Key West, Monroe County   fishing   
3691                                     Bel Air, Dakar    diving   
4058                                     Astwood's Cove       nan   
4101                                    Port Elizabeth    fishing   
5321                     St. Augustine, St Johns County       nan   
6108                                                NaN       nan   
6147                                         St. Tropez   bathing   

                           name  sex  age  \
119                        male    M  NaN   
290             Jonathan Lester    M   34   
382                        male    M   40   
631               Bryan Collins    M  NaN   
1130                     Alicia    F  NaN   
1242               Kobus Maritz    M   46   
2125              Sylvia Lanner    F   41   
2144                       male    M  NaN   
2460               Jeff Weakley    M  NaN   
2648                       male    M   17   
3508                        NaN  NaN  NaN   
3691                       O.D.    M  NaN   
4058  Dorothy Barbara Rawlinson    F   29   
4101                       male    M  NaN   
5321            M.F. Turnipseed    M  NaN   
6108                       male  NaN  NaN   
6147            Sam

#### Comprobación tipo de dato de la columna `date`

In [75]:
dc.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6302 entries, 0 to 6301
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   casenumber            6301 non-null   object 
 1   date                  6302 non-null   object 
 2   year                  6300 non-null   float64
 3   type                  6298 non-null   object 
 4   country               6302 non-null   object 
 5   area                  5847 non-null   object 
 6   location              5762 non-null   object 
 7   activity              6302 non-null   object 
 8   name                  6092 non-null   object 
 9   sex                   5737 non-null   object 
 10  age                   3471 non-null   object 
 11  injury                6274 non-null   object 
 12  fatal(y/n)            5763 non-null   object 
 13  time                  2948 non-null   object 
 14  species               6302 non-null   object 
 15  investigatororsource  6302

In [76]:
type(dc.date[0]) # Comprobamos que es de tipo string/objeto -> Vamos a cambiarlo a datetime

str

In [77]:
# Para poder pasar los datos a datetime, primero voy a quitar la palabra reported de aquellas casillas en las que pone'Reported en x fecha'
dc.date = dc.date.str.replace('Reported','')

In [78]:
# Voy a tratar de limpiar la columna en un solo paso y pasando a nan los datos que no coincidan con el formato dd-mmm-yyyy
dc.date = pd.to_datetime(dc.date, format='%d-%b-%Y',errors='coerce')

In [79]:
dc[dc.date.notna()].shape

(4765, 24)

In [80]:
# Voy a borrar todas aquellas entradas que no tengan el formato de fecha adecuado
indices_fecha_mala = dc[dc.date.isna()].index
indices_fecha_mala[:10]

Index([14, 17, 31, 59, 62, 65, 86, 90, 110, 122], dtype='int64')

In [81]:
dc.drop(index=indices_fecha_mala, inplace=True)

In [113]:
dc.reset_index().head()

index  casenumber       date  year        type    country             area  \
0      0  2018.06.25 2018-06-25  2018     Boating        USA       California   
1      1  2018.06.18 2018-06-18  2018  Unprovoked        USA          Georgia   
2      2  2018.06.09 2018-06-09  2018     Invalid        USA           Hawaii   
3      3  2018.06.08 2018-06-08  2018  Unprovoked  AUSTRALIA  New South Wales   
4      4  2018.06.04 2018-06-04  2018    Provoked     MEXICO           Colima   

                         location  activity             name sex  age  \
0     Oceanside, San Diego County  paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County  standing  Adyson McNeely    F   11   
2                    Habush, Oahu   surfing      John Denges   M   48   
3              Arrawarra Headland   surfing             male   M  NaN   
4                        La Ticla    diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00          unknown  K.McMurray, TrackingSharks.com   
2          07h45          unknown  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

In [83]:
dc.shape

(4765, 24)

In [84]:
# Se ha reducido la base de datos a 4769 filas
dc.head(10)

casenumber       date    year        type    country             area  \
0    2018.06.25 2018-06-25  2018.0     Boating        USA       California   
1    2018.06.18 2018-06-18  2018.0  Unprovoked        USA          Georgia   
2    2018.06.09 2018-06-09  2018.0     Invalid        USA           Hawaii   
3    2018.06.08 2018-06-08  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4    2018.06.04 2018-06-04  2018.0    Provoked     MEXICO           Colima   
5  2018.06.03.b 2018-06-03  2018.0  Unprovoked  AUSTRALIA  New South Wales   
6  2018.06.03.a 2018-06-03  2018.0  Unprovoked     BRAZIL       Pernambuco   
7    2018.05.27 2018-05-27  2018.0  Unprovoked        USA          Florida   
8  2018.05.26.b 2018-05-26  2018.0  Unprovoked        USA          Florida   
9  2018.05.26.a 2018-05-26  2018.0  Unprovoked        USA          Florida   

                                            location  activity  \
0                        Oceanside, San Diego County  paddling   
1                     St. Simon Island, Glynn County  standing   
2                                       Habush, Oahu   surfing   
3                                 Arrawarra Headland   surfing   
4                                           La Ticla    diving   
5                                 Flat Rock, Ballina   surfing   
6                              Piedade Beach, Recife  swimming   
7  Lighhouse Point Park, Ponce Inlet, Volusia County   fishing   
8                       Cocoa Beach, Brevard  County   walking   
9                      Daytona Beach, Volusia County  standing   

                     name sex  age  \
0             Julie Wolfe   F   57   
1         Adyson McNeely    F   11   
2             John Denges   M   48   
3                    male   M  NaN   
4          Gustavo Ramos    M  NaN   
5                 Chris 
   M  NaN   
6  Jose Ernesto da Silva    M   18   
7                    male   M   52   
8               Cody High   M   15   
9                    male   M   12   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   
5                            No injury, board bitten          N   
6                                              FATAL          Y   
7            Minor injury to foot. PROVOKED INCIDENT          N   
8                              Lower left leg bitten          N   
9                               Minor injury to foot          N   

             time          species              investigatororsource  \
0           18h00      White shark                  R. Collier, GSAF   
1   14h00  -15h00          unknown    K.McMurray, TrackingSharks.com   
2           07h45          unknown    K.McMurray, TrackingSharks.com   
3             NaN        2 m shark                    B. Myatt, GSAF   
4             NaN  Tiger shark, 3m                         A .Kipper   
5             NaN          unknown         Daily Telegraph, 6/4/2018   
6  Late afternoon      Tiger shark    Diario de Pernambuco, 6/4/2018   
7             NaN  Lemon shark, 3'   K. McMurray, TrackingSharks.com   
8           17h00   Bull shark, 6'    K.McMurray, TrackingSharks.com   
9           14h00          unknown  K. McMurray, Tracking Sharks.com   

                             pdf  \
0           2018.06.25-Wolfe.pdf   
1         2018.06.18-McNeely.pdf   
2          2018.06.09-Denges.pdf   
3       2018.06.08-Arrawarra.pdf   
4           2018.06.04-Ramos.pdf   
5      2018.06.03.b-FlatRock.pdf   
6       2018.06.03.a-daSilva.pdf   
7           2018.05.27-Ponce.pdf   
8          2018.05.26.b-High.pdf   
9  2018.05.26.a-DaytonaBeach.pdf   

                                         hrefformula  \
0  http://sharka

In [85]:
dc.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4765 entries, 0 to 6160
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   casenumber            4765 non-null   object        
 1   date                  4765 non-null   datetime64[ns]
 2   year                  4765 non-null   float64       
 3   type                  4763 non-null   object        
 4   country               4765 non-null   object        
 5   area                  4564 non-null   object        
 6   location              4540 non-null   object        
 7   activity              4765 non-null   object        
 8   name                  4649 non-null   object        
 9   sex                   4415 non-null   object        
 10  age                   3138 non-null   object        
 11  injury                4750 non-null   object        
 12  fatal(y/n)            4375 non-null   object        
 13  time                  2

In [86]:
# Voy a intentar cambiar tb la columna year
dc.year = dc.year.astype(int)

In [87]:
dc.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4765 entries, 0 to 6160
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   casenumber            4765 non-null   object        
 1   date                  4765 non-null   datetime64[ns]
 2   year                  4765 non-null   int32         
 3   type                  4763 non-null   object        
 4   country               4765 non-null   object        
 5   area                  4564 non-null   object        
 6   location              4540 non-null   object        
 7   activity              4765 non-null   object        
 8   name                  4649 non-null   object        
 9   sex                   4415 non-null   object        
 10  age                   3138 non-null   object        
 11  injury                4750 non-null   object        
 12  fatal(y/n)            4375 non-null   object        
 13  time                  2

In [88]:
dc.shape

(4765, 24)

#### Columna `fatal(y/n)`
* En esta columna voy a rellenar los nans con N a no ser que se detecte la palabra fatal en la columna `injury`
* El resto de errores los voy a solucionar uno a uno

In [89]:
def simuerto(cadena): # Defino una función que devuelva a la columna fatal(y/n) N si la palabra fatal no se lee o una Y si sí se encuentra la misma
    if 'fatal' in str(cadena).lower(): # Si se encuentra la palabra
        return 'Y'
    else: 
        return 'N'

In [90]:
dc['fatal(y/n)'].value_counts()

fatal(y/n)
N          3438
Y           891
UNKNOWN      39
 N            5
M             1
2017          1
Name: count, dtype: int64

In [91]:
indices_fatalmala = dc['fatal(y/n)'][dc['fatal(y/n)'].isna()].index
indices_fatalmala.shape

(390,)

In [92]:
dc.loc[indices_fatalmala].head(10) # Muestro un head de las columnas que tengan valores nan -> Solo quiero operar la funcion definida sobre estas filas

casenumber       date  year        type       country  \
27   2018.04.15.c 2018-04-15  2018  Unprovoked        BRAZIL   
76     2017.10.01 2017-10-01  2017     Invalid  SOUTH AFRICA   
79   2017.09.24.b 2017-09-24  2017     Invalid           USA   
84   2017.09.15.b 2017-09-15  2017     Invalid  SOUTH AFRICA   
98   2017.08.26.b 2017-08-26  2017     Invalid         SPAIN   
107  2017.07.29.c 2017-07-29  2017     Invalid           USA   
115  2017.07.20.c 2017-07-20  2017     Invalid           USA   
124    2017.07.02 2017-07-02  2017     Invalid       COMOROS   
125    2017.06.29 2017-06-29  2017     Invalid           USA   
145    2017.05.03 2017-05-03  2017     Invalid           USA   

                      area                             location  activity  \
27              Pernambuco                              Piedade  swimming   
76   Western Cape Province                          Dyer Island    diving   
79                New York                             Rockaway   surfing   
84   Western Cape Province                              Hawston    diving   
98               Castellón                      Grao de Moncofa  swimming   
107         South Carolina  Hilton Head Island, Beaufort County  swimming   
115         South Carolina  Hilton Head Island, Beaufort County  swimming   
124                Anjouan                                 Moya   fishing   
125         South Carolina                   Wrightsville Beach   fishing   
145             California          Sunset Beach, Orange County   surfing   

                   name  sex  age  \
27        Pablo de Melo    M   34   
76         Bradley Fick    M   31   
79       Michah Behrend    M   33   
84           Wayon Love    M   25   
98               female    F   11   
107  Johnny Simatacolos    M   10   
115        Ellie Rogier    F    8   
124                 NaN  NaN  NaN   
125                male    M  NaN   
145         Sophia Raab    F   18   

                                                injury fatal(y/n)       time  \
27   Multiple severe injuries to arms and leg, leg ...        NaN        NaN   
76                                               FATAL        NaN        NaN   
79              Lacerations to right ankle, foot & toe        NaN      15h30   
84       FATAL, but death was probably due to drowning        NaN  Afternoon   
98                            Lacerations to left foot        NaN     Midday   
107                 Abrasions and cuts to sole of foot        NaN      15h00   
115                                       Foot injured        NaN        NaN   
124  Skull found in shark, a probable drowning & sc...        NaN        NaN   
125                Arm injured by hook, not by a shark        NaN      19h30   
145  Laceration to thigh, likely caused by surfboar...        NaN      14h30   

                                            species  \
27                                          unknown   
76              Death may have been due to drowning   
79                                     Questionable   
84                                          unknown   
98                   Shark involvement questionable   
107                                   A small shark   
115                 Shark involvement not confirmed   
124  Shark involvement prior to death not confirmed   
125                           No shark invovlement    
145               Shark involvement highly doubtful   

                     investigatororsource                           pdf  \
27                       Globo, 4/16/2018       2018.04.15.c-deMelo.pdf   
76                All Africa, 10/11/2017            2017.10.01-Fick.pdf   
79              S. Curatolo-Wageman, GSAF      2017.09.24.b-Behrend.pdf   
84                   Ground Up, 9/20/2017         2017.09.15.b-Love.pdf   
98   El Periodico Mediterraneo, 8/27/2017        2017.08.26.b-Spain.pdf   
107                     C. Creswell, GSAF  2017.07.29.c-Simatacolos.pdf   
115                     C. Creswel

In [114]:
# Procedo a limpiar la columna de valores nulos:
dc.loc[indices_fatalmala, 'injury'].apply(simuerto).head()


27    N
76    Y
79    N
84    Y
98    N
Name: injury, dtype: object

In [94]:
dc.loc[indices_fatalmala, 'fatal(y/n)'] = dc.loc[indices_fatalmala, 'injury'].apply(simuerto) # Aplicamos cambios al dataframe

In [95]:
dc.loc[indices_fatalmala].head(5) # Comprobamos que se ha hecho bien -> Parece que sí

casenumber       date  year        type       country  \
27  2018.04.15.c 2018-04-15  2018  Unprovoked        BRAZIL   
76    2017.10.01 2017-10-01  2017     Invalid  SOUTH AFRICA   
79  2017.09.24.b 2017-09-24  2017     Invalid           USA   
84  2017.09.15.b 2017-09-15  2017     Invalid  SOUTH AFRICA   
98  2017.08.26.b 2017-08-26  2017     Invalid         SPAIN   

                     area         location  activity            name sex age  \
27             Pernambuco          Piedade  swimming   Pablo de Melo   M  34   
76  Western Cape Province      Dyer Island    diving    Bradley Fick   M  31   
79               New York         Rockaway   surfing  Michah Behrend   M  33   
84  Western Cape Province          Hawston    diving      Wayon Love   M  25   
98              Castellón  Grao de Moncofa  swimming          female   F  11   

                                               injury fatal(y/n)       time  \
27  Multiple severe injuries to arms and leg, leg ...          N        NaN   
76                                              FATAL          Y        NaN   
79             Lacerations to right ankle, foot & toe          N      15h30   
84      FATAL, but death was probably due to drowning          Y  Afternoon   
98                           Lacerations to left foot          N     Midday   

                                species                  investigatororsource  \
27                              unknown                      Globo, 4/16/2018   
76  Death may have been due to drowning               All Africa, 10/11/2017    
79                         Questionable             S. Curatolo-Wageman, GSAF   
84                              unknown                  Ground Up, 9/20/2017   
98       Shark involvement questionable  El Periodico Mediterraneo, 8/27/2017   

                         pdf  \
27   2018.04.15.c-deMelo.pdf   
76       2017.10.01-Fick.pdf   
79  2017.09.24.b-Behrend.pdf   
84     2017.09.15.b-Love.pdf   
98    2017.08.26.b-Spain.pdf   

                                          hrefformula  \
27  http://sharkattackfile.net/spreadsheets/pdf_di...   
76  http://sharkattackfile.net/spreadsheets/pdf_di...   
79  http://sharkattackfile.net/spreadsheets/pdf_di...   
84  http://sharkattackfile.net/spreadsheets/pdf_di...   
98  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                 href  casenumber.1  \
27  http://sharkattackfile.net/spreadsheets/pdf_di...  2018.04.15.c   
76  http://sharkattackfile.net/spreadsheets/pdf_di...    2017.10.01   
79  http://sharkattackfile.net/spreadsheets/pdf_di...  2017.09.24.b   
84  http://sharkattackfile.net/spreadsheets/pdf_di...  2017.09.15.b   
98  http://sharkattackfile.net/spreadsheets/pdf_di...  2017.08.26.b   

    casenumber.2  originalorder unnamed:22 unnamed:23  
27  2018.04.15.c         6276.0        NaN        NaN  
76    2017.10.01         6227.0        NaN        NaN  
79  2017.09.24.b         6224.0        NaN        NaN  
84  2017.09.15.b         6219.0        NaN        NaN  
98  2017.08.26.b         6205.0        NaN        NaN

In [96]:
dc['fatal(y/n)'].value_counts()

fatal(y/n)
N          3793
Y           926
UNKNOWN      39
 N            5
M             1
2017          1
Name: count, dtype: int64

In [97]:
# Vamos a quitar los espacios de la N
dc['fatal(y/n)'] = dc['fatal(y/n)'].str.replace(' ', '')

In [98]:
dc['fatal(y/n)'].value_counts()

fatal(y/n)
N          3798
Y           926
UNKNOWN      39
M             1
2017          1
Name: count, dtype: int64

In [99]:
dc['fatal(y/n)'].value_counts()

fatal(y/n)
N          3798
Y           926
UNKNOWN      39
M             1
2017          1
Name: count, dtype: int64

In [100]:
# Vamos a aplicar la función de ismuerto() también a los datos que tengan un unknown. Procedemos igual que antes -> Sacamos indices y le aplicamos la función a ese rango
indices_fatal_unkown = dc['fatal(y/n)'][dc['fatal(y/n)'] == 'UNKNOWN'].index
indices_fatal_unkown.shape

(39,)

In [101]:
dc.loc[indices_fatal_unkown, 'fatal(y/n)'] = dc.loc[indices_fatal_unkown, 'injury'].apply(simuerto) # Aplicamos cambios al dataframe

In [102]:
dc.loc[indices_fatal_unkown].head(5) # Comprobamos haberlo hecho bien

casenumber       date  year        type    country             area  \
624     2013.09.08 2013-09-08  2013  Unprovoked        USA   South Carolina   
1270  2008.04.20.a 2008-04-20  2008  Unprovoked  AUSTRALIA  New South Wales   
2250    1997.06.07 1997-06-07  1997  Unprovoked     BRAZIL   Rio de Janeiro   
2260    1997.02.21 1997-02-21  1997  Unprovoked        USA           Hawaii   
2267    1996.12.29 1996-12-29  1996  Unprovoked  AUSTRALIA       Queensland   

                                location activity               name sex  age  \
624   St. Helena Island, Beaufort County      nan             female   F  NaN   
1270                       Crescent Head      nan    Jamie Adlington   M  NaN   
2250          Copacabana, Rio de Janeiro  bathing  José Luiz Lipiani   M  NaN   
2260                 Sunset Beach, O'ahu      nan    Gersome Perreno   M  NaN   
2267                        Coolum Beach  surfing         Blair Hall   M   18   

          injury fatal(y/n)   time             species  \
624   No details          N    NaN             unknown   
1270         NaN          N    NaN  Tiger shark, 2.3m    
2250         NaN          N    NaN             unknown   
2260  No details          N    NaN             unknown   
2267         NaN          N  18h00             unknown   

                 investigatororsource                         pdf  \
624                  WIS-TV, 9/9/2013    2013.09.08-St-Helena.pdf   
1270                   T. Peake, GSAF  2008.04.20.a-Adlington.pdf   
2250                  Globo, 6/9/1997   1997.06.07-NV-Lipiani.pdf   
2260                        G. Balazs   1997.02.21-NV-Perreno.pdf   
2267  The Advertiser, 12/30/1996, p.3    1996.12.29-BlairHall.pdf   

                                            hrefformula  \
624   http://sharkattackfile.net/spreadsheets/pdf_di...   
1270  http://sharkattackfile.net/spreadsheets/pdf_di...   
2250  http://sharkattackfile.net/spreadsheets/pdf_di...   
2260  http://sharkattackfile.net/spreadsheets/pdf_di...   
2267  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href  casenumber.1  \
624   http://sharkattackfile.net/spreadsheets/pdf_di...    2013.09.08   
1270  http://sharkattackfile.net/spreadsheets/pdf_di...  2008.04.20.a   
2250  http://sharkattackfile.net/spreadsheets/pdf_di...    1997.06.07   
2260  http://sharkattackfile.net/spreadsheets/pdf_di...    1997.02.21   
2267  http://sharkattackfile.net/spreadsheets/pdf_di...    1996.12.29   

      casenumber.2  originalorder unnamed:22 unnamed:23  
624     2013.09.08         5679.0        NaN        NaN  
1270  2008.04.20.a         5033.0        NaN        NaN  
2250    1997.06.07         4053.0        NaN        NaN  
2260    1997.02.21         4043.0        NaN        NaN  
2267    1996.12.29         4036.0        NaN        NaN

In [103]:
dc['fatal(y/n)'].value_counts()

fatal(y/n)
N       3837
Y        926
M          1
2017       1
Name: count, dtype: int64

In [104]:
# Vamos a cambiar la M por una N asumiendo que es una errata (las dos teclas se encuentran juntas) y a aplicarle la fórmula de simuerto() a la fila del '2017'
dc['fatal(y/n)'] = dc['fatal(y/n)'].str.replace('M', 'N')

In [105]:
indices_fatal_17 = dc['fatal(y/n)'][dc['fatal(y/n)'] == '2017'].index
indices_fatal_17.shape

(1,)

In [106]:
dc.loc[indices_fatal_17, 'fatal(y/n)'] = dc.loc[indices_fatal_17, 'injury'].apply(simuerto) # Aplicamos cambios al dataframe

In [107]:
dc['fatal(y/n)'].value_counts()

fatal(y/n)
N    3839
Y     926
Name: count, dtype: int64

In [108]:
# Hecho!

Llegados a este punto vamos a guardar esta base de datos como .csv y pasar a trabajar en el análisis de la misma en otro fichero llamado analysis.ipynb

#### Exportado base de datos

In [109]:
dc.to_csv('../data/attacks_limpio.csv', index=False)